source/extensions/clusters/redis/redis_cluster.cc

```diff
@@ -0,0 +1,278 @@
+
+RedisCluster::RedisDiscoverySession::~RedisDiscoverySession() {
+  if (current_request_) {
+    current_request_->cancel();
+    current_request_ = nullptr;
+  }
+
+  while (!client_map_.empty()) {
+    client_map_.begin()->second->client_->close();
+  }
+}
+
+void RedisCluster::RedisDiscoveryClient::onEvent(Network::ConnectionEvent event) {
+  if (event == Network::ConnectionEvent::RemoteClose ||
+      event == Network::ConnectionEvent::LocalClose) {
+    auto client_to_delete = parent_.client_map_.find(host_);
+    ASSERT(client_to_delete != parent_.client_map_.end());
+    parent_.dispatcher_.deferredDelete(std::move(client_to_delete->second->client_));
+    parent_.client_map_.erase(client_to_delete);
+  }
+}
+
+void RedisCluster::RedisDiscoverySession::registerDiscoveryAddress(
+    const std::list<Envoy::Network::Address::InstanceConstSharedPtr>& address_list,
+    const uint32_t port) {
+  // Since the address from DNS does not have port, we need to make a new address that has port in
+  // it.
+  for (const Network::Address::InstanceConstSharedPtr& address : address_list) {
+    ASSERT(address != nullptr);
+    discovery_address_list_.push_back(Network::Utility::getAddressWithPort(*address, port));
+  }
+}
+
+void RedisCluster::RedisDiscoverySession::startResolve() {
+  parent_.info_->stats().update_attempt_.inc();
+  // If a resolution is currently in progress, skip it.
+  if (current_request_) {
+    return;
+  }
+
+  // If hosts is empty, we haven't received a successful result from the CLUSTER SLOTS call yet.
+  // So, pick a random discovery address from dns and make a request.
+  Upstream::HostSharedPtr host;
+  if (hosts_.empty()) {
+    const int rand_idx = parent_.random_.random() % discovery_address_list_.size();
+    auto it = discovery_address_list_.begin();
+    std::next(it, rand_idx);
+    host = Upstream::HostSharedPtr{new RedisHost(parent_.info(), "", *it, parent_, true)};
+  } else {
+    const int rand_idx = parent_.random_.random() % hosts_.size();
+    host = hosts_[rand_idx];
+  }
+
+  current_host_address_ = host->address()->asString();
+  RedisDiscoveryClientPtr& client = client_map_[current_host_address_];
+  if (!client) {
+    client = std::make_unique<RedisDiscoveryClient>(*this);
+    client->host_ = current_host_address_;
+    client->client_ = client_factory_.create(host, dispatcher_, *this);
+    client->client_->addConnectionCallbacks(*client);
+  }
+
+  current_request_ = client->client_->makeRequest(ClusterSlotsRequest::instance_, *this);
+}
+
+void RedisCluster::RedisDiscoverySession::onResponse(
+    NetworkFilters::Common::Redis::RespValuePtr&& value) {
+  current_request_ = nullptr;
+
+  if (value->type() != NetworkFilters::Common::Redis::RespType::Array) {
+    ENVOY_LOG(warn, "Unexpected response to cluster slot command: {}", value->toString());
+    parent_.info_->stats().update_no_rebuild_.inc();
+    return;
+  }
+
+  Upstream::HostVector new_hosts;
+  std::array<std::string, 16384> slots_;
```

Comment:
I know there has been some discussion here of how we are doing this, but is it possible to put a description of the algorithm somewhere for a non-Redis Cluster expert? Presumably this is a data structure that we will end up eventually load balancing on to? Is that right? Once we start load balancing, is this going to have to use TLS so that we can do per-worker load balancing on top of this slot structure?

Are there any practical alternatives to the fixed size array? I'm just trying to understand the tradeoffs.

Reply:
I think the primary concern is the repeated copy of std::string.  I've updated the code so that we store the array of shared_ptr to the host object.  I don't think we'll have to use TLS, we can just use this slot structure directly.  The main advantage of the fixed size array is that look up is extremely fast.  The memory used by this data structure is 130KB per cluster, I think this is a reasonable tradeoff.